cloudevents/working-drafts/minor-versions.md

```diff
@@ -0,0 +1,337 @@
+evt["xyz"] = "test";
+
+// Likewise accessing the property for an event parsed from elsewhere:
+evt = httpRequest.ParseCloudEvent();
+// If evt is v1.0, this should fail *even if it has an xyz attribute value that's valid in v1.1.
+int xyz = evt.Xyz;
+```
+
+Setting a value within an event where the version indicates that the attribute *is* known should
+validate the attribute value according to that version. Example:
+
+```csharp
+// Create a v1.1 event
+var evt = new CloudEvent(CloudEventsSpecVersion.V1_1);
+
+// This should work: 1 is a valid value, and xyz is known in v1.1
+evt.Xyz = 1;
+
+// This should fail: 123 is not a valid value for xyz in v1.1
+evt.Xyz = 123;
+
+// This should fail to compile, where that's feasible: xyz should always be an integer
+evt.Xyz = "test";
+
+// This should succeed at execution time
+evt["xyz"] = 123;
+
+// This should fail at execution time due to validation
+evt["xyz"] = "test";
+
+// This should fail at execution time due to validation
+evt["xyz"] = 123;
+```
+
+### "Changing" event versions
+
+An SDK may provide the ability to change the specversion of an event,
+or (preferred) create a new object based on the old one. When "upgrading" an event's specversion,
+the SDK should perform validation as if parsing the event from scratch. Examples:
+
+```csharp
+// Successful upgrade: no xyz value
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0) { Id = "..." };
+var evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Successful upgrade: valid value for xyz
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0)
+{
+    Id = "...",
+    ["xyz"] = 10
+};
+evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Failed upgrade: invalid value for xyz
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0)
+{
+    Id = "...",
+    ["xyz"] = "test"
+};
+evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Successful downgrade
+var evt11 = new CloudEvent(CloudEventsSpecVersion.V1_1)
+{
+    Id = "...",
+    Xyz = 10
+};
+evt10 = evt11.WithSpecVersion(CloudEventsSpecVersion.V1_0);
+// Access to evt10.Xyz would fail as described above;
+// access to evt10["xyz"] would return 10.
+// Setting evt10["xyz"] to "test" or 123 could fail or succeed - it's
+// up to the SDK to decide (and document!) whether downgrading retains constraints
+// for attributes that were present in the original event.
+// TODO: What about if Xyz hadn't been present before? Should evt10["xyz"] = "test" fail then?
+```
+
+Until we have a new major version, we probably don't want to specify whether
+SDKs should allow upgrade/downgrade across major versions.
+
+### Handling unknown versions
+
+An SDK which supports minor versions **MUST** be able to parse events with
+a minor version of which it's unaware, but **MUST NOT** parse events with a
+*major* version of which it's unaware. In other words, a v1.1-aware SDK must
+be able to parse a v1.2 event, but must reject any attempt to parse a v2.0 event.
```

Comment:
Would the resulting CE be a 1.1 CE or a 1.2 CE?
No matter which way it goes something feels wrong.
If it's 1.1 then are we lying to the user because the CE came in as 1.2.
If it's 1.2 then are we teasing the user because they can't access the 1.2 attributes as first class attributes and we won't do any validation on the 1.2 attributes like we would for 1.1 attributes?

Reply:
It would be a 1.2 event, but I'd expect the SDK to have a way of saying "This is 1.2 and I don't understand that fully."

Reply:
At the moment the go sdk fails if you present a minor version it does not understand, but I could see how we could support the idea of a typed "Next" version in the same minor family. It would have core members that match the newest known version but present newer core members as extensions to the integration. It would also possibly change some validation rules, mostly about what a valid version number is.